cache http probe request object

Taoup opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Http probe requests are constructed for every http probe that's being executed, which is probably a waste of cpu cycles and memory. This PR uses sync.Map as a cache to reuse the constructed http requests.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #115939 

#### Special notes for your reviewer:
The original PR #116058 has been closed and not active for quite some time, so I'd like to present my implementation for discussion. 

Besides, I'm new to Kubernetes and the golang overall.  I'd like to submit is PR asking for help  to check the sanity of my implementation before moving on to next step, i.e. benchmarking and so on. 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Taoup committed:
cache http probe request with sync.Map

Taoup commented:
/retest

haircommander commented:
/priority important-longterm
/triage accepted
/release-note-none

Taoup committed:
benchmark http probe w/o http request cache

cache stored in prober struct, reusing probeKey as cache key.

goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/kubelet/prober
cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
            │    old.txt    │               new.txt               │
            │    sec/op     │   sec/op     vs base                │
HTTPProbe-8   142.71m ± 10%   33.33m ± 5%  -76.65% (p=0.000 n=10)

            │    old.txt    │               new.txt                │
            │     B/op      │     B/op      vs base                │
HTTPProbe-8   251.43Mi ± 0%   35.01Mi ± 0%  -86.08% (p=0.000 n=10)

            │   old.txt    │               new.txt               │
            │  allocs/op   │  allocs/op   vs base                │
HTTPProbe-8   3367.7k ± 0%   607.0k ± 0%  -81.98% (p=0.000 n=10)

Taoup commented:
update benchmark result:

```
benchstat old.txt new.txt 
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/kubelet/prober
cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
        │   old.txt   │               new.txt               │
        │   sec/op    │   sec/op     vs base                │
Probe-8   557.6m ± 5%   164.7m ± 6%  -70.47% (p=0.000 n=10)

        │   old.txt    │               new.txt                │
        │     B/op     │     B/op      vs base                │
Probe-8   359.9Mi ± 0%   142.2Mi ± 0%  -60.49% (p=0.000 n=10)

        │   old.txt    │               new.txt               │
        │  allocs/op   │  allocs/op   vs base                │
Probe-8   3466.0k ± 0%   693.2k ± 0%  -80.00% (p=0.000 n=10)
```

Taoup commented:
> update benchmark result:
> 
> ```
> benchstat old.txt new.txt 
> goos: linux
> goarch: amd64
> pkg: k8s.io/kubernetes/pkg/kubelet/prober
> cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
>         │   old.txt   │               new.txt               │
>         │   sec/op    │   sec/op     vs base                │
> Probe-8   557.6m ± 5%   164.7m ± 6%  -70.47% (p=0.000 n=10)
> 
>         │   old.txt    │               new.txt                │
>         │     B/op     │     B/op      vs base                │
> Probe-8   359.9Mi ± 0%   142.2Mi ± 0%  -60.49% (p=0.000 n=10)
> 
>         │   old.txt    │               new.txt               │
>         │  allocs/op   │  allocs/op   vs base                │
> Probe-8   3466.0k ± 0%   693.2k ± 0%  -80.00% (p=0.000 n=10)
> ```

This benchmark does probes in serial mode, which I thought is way more easier to implement than a parallel one. I'm not sure if this result is representative enough. Require for more comments, :)

Taoup commented:
Update: parallel benchmark result
```
root ➜ /workspaces/kubernetes (master) $ benchstat old.txt new.txt 
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/kubelet/prober
cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
        │    old.txt    │               new.txt                │
        │    sec/op     │    sec/op     vs base                │
Probe-8   173.10m ± 19%   21.70m ± 19%  -87.46% (p=0.000 n=10)

        │    old.txt    │               new.txt                │
        │     B/op      │     B/op      vs base                │
Probe-8   251.30Mi ± 0%   33.66Mi ± 0%  -86.61% (p=0.000 n=10)

        │   old.txt    │               new.txt               │
        │  allocs/op   │  allocs/op   vs base                │
Probe-8   3367.5k ± 0%   594.7k ± 0%  -82.34% (p=0.000 n=10)
```

the code doing this benchmark:
https://github.com/Taoup/kubernetes/commit/853845791b5f8ab1ce36ff113f9e433f2ec4c343

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -142,9 +147,27 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
+		}
+		// cache hit, but pod IP changed, recreate the request object
+		if p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP {
+			req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
+			if err != nil {
+				return probe.Unknown, "", err
+			}
+			httpProbeReqCache.Store(key, req)
```
Do I understand correctly that in case of cache miss this is repeated twice?
```
req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
if err != nil {
	return probe.Unknown, "", err
}
httpProbeReqCache.Store(key, req)
```

Would it make sense to combine `!found` and this scope?

Taoup replied:
```
Would it make sense to combine !found and this scope?
```
Done

aojea commented on pkg/kubelet/prober/prober.go:
```diff
@@ -40,6 +42,9 @@ import (
 const maxProbeRetries = 3
 
+// Requests cache for http probe
+var httpProbeReqCache sync.Map
```
why a global and not part of the prober struct?

Taoup replied:
I tried to move the cache to httpprober. However, that prevent me from reusing the probeKey struct in the prober package. While move the cache to prober struct is quite simple and straightforward.

aojea replied:
we have bad experience caching globally because they can create memory leaks if some objects are never deleted , and those bugs were very subtle to catch. I think we should avoid creating a cache that is global

bart0sh replied:
> I tried to move the cache to httpprober. However, that prevent me from reusing the probeKey struct in the prober package. While move the cache to prober struct is quite simple and straightforward.

Well, as caching makes sense only for HTTP probes, the logical place for it is the `httpprober` from my point of view. Reusing probeKey shouldn't stop you from at least trying. You can either move it somewhere (to avoid cyclical import) and make it public, or even duplicating in in the `httpprober`, which is also ok from my point of view.

I'd suggest to look at the possibility to cache http requests in the`httpprobe.NewRequestForHTTPGetAction` and may be renaming this method. This would allow you to cache [lifecycle probes](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/lifecycle/handlers.go#L159) as well.

Taoup replied:
> Well, as caching makes sense only for HTTP probes, the logical place for it is the httpprober from my point of view. Reusing probeKey shouldn't stop you from at least trying. You can either move it somewhere (to avoid cyclical import) and make it public, or even duplicating in in the httpprober, which is also ok from my point of view.

I agree that the logical place for the cache should be in httpprober. If you prefer, I could implement the cache in httpprober to make a comparison.

> I'd suggest to look at the possibility to cache http requests in thehttpprobe.NewRequestForHTTPGetAction and may be renaming this method. This would allow you to cache [lifecycle probes](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/lifecycle/handlers.go#L159) as well.

I read through the code and the docs, it seems that lifecycle hooks does not  get executed as frequently as normal probes, i.e. only upon container start and termination. I guess it does't worth a cache to do that.

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -183,6 +197,13 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
 }
 
+func cachedRequestStaled(p *v1.Probe, req *http.Request, status v1.PodStatus) bool {
+	if p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP {
+		return true
+	}
+	return false
```
```suggestion
	return p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP {
```

Considering the suggestion, do we need a separate function for this?

Taoup replied:
A separate function would help with readability and make the original if statement not too verbose. without it, it will look like this.
```go
		reqAny, found := pb.httpProbeReqCache.Load(key)
		if !found || (p.HTTPGet.Host == "" && reqAny.(*http.Request).URL.Hostname() != status.PodIP) {
			req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
			if err != nil {
```

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -142,9 +147,18 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
+		var req *http.Request
+		key := probeKey{podUID: pod.UID, containerName: container.Name, probeType: probeType}
+		reqAny, found := httpProbeReqCache.Load(key)
+		if !found || cachedRequestStaled(p, reqAny.(*http.Request), status) {
+			req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
+			if err != nil {
+				return probe.Unknown, "", err
+			}
+			httpProbeReqCache.Store(key, req)
+		} else {
+			req = reqAny.(*http.Request)
 		}
```
```suggestion
                req = reqAny.(*http.Request)
		if !found || (p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP) {
			req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
			if err != nil {
				return probe.Unknown, "", err
			}
			httpProbeReqCache.Store(key, req)
		}
```

Taoup replied:
Sorry, as a newbie in go, I'm not sure what will happen in case of a cache miss, reqAny will be nil? Is it ok to do a type assertion in that case?

bart0sh replied:
Good question. Go will panic trying to convert nil interface to *http.Request, I guess. My bad.

bart0sh replied:
I've updated the suggestion.

Taoup replied:
this is really neat. I'll make those changes

bart0sh replied:
Frankly, It doesn't look very readable to me. I'd even try to avoid using sync.Map because of this and use plain map + locks (if locks are really needed here).

aojea commented:
> update benchmark result:
> 
> ```
> benchstat old.txt new.txt 
> goos: linux
> goarch: amd64
> pkg: k8s.io/kubernetes/pkg/kubelet/prober
> cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
>         │   old.txt   │               new.txt               │
>         │   sec/op    │   sec/op     vs base                │
> Probe-8   557.6m ± 5%   164.7m ± 6%  -70.47% (p=0.000 n=10)
> 
>         │   old.txt    │               new.txt                │
>         │     B/op     │     B/op      vs base                │
> Probe-8   359.9Mi ± 0%   142.2Mi ± 0%  -60.49% (p=0.000 n=10)
> 
>         │   old.txt    │               new.txt               │
>         │  allocs/op   │  allocs/op   vs base                │
> Probe-8   3466.0k ± 0%   693.2k ± 0%  -80.00% (p=0.000 n=10)
> ```

can you add the benchmark to the code and reflect the results in a comment?

Taoup commented:
> > update benchmark result:
> > ```
> > benchstat old.txt new.txt 
> > goos: linux
> > goarch: amd64
> > pkg: k8s.io/kubernetes/pkg/kubelet/prober
> > cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
> >         │   old.txt   │               new.txt               │
> >         │   sec/op    │   sec/op     vs base                │
> > Probe-8   557.6m ± 5%   164.7m ± 6%  -70.47% (p=0.000 n=10)
> > 
> >         │   old.txt    │               new.txt                │
> >         │     B/op     │     B/op      vs base                │
> > Probe-8   359.9Mi ± 0%   142.2Mi ± 0%  -60.49% (p=0.000 n=10)
> > 
> >         │   old.txt    │               new.txt               │
> >         │  allocs/op   │  allocs/op   vs base                │
> > Probe-8   3466.0k ± 0%   693.2k ± 0%  -80.00% (p=0.000 n=10)
> > ```
> 
> can you add the benchmark to the code and reflect the results in a comment?


the code doing this benchmark: 4c0578c9c8e19ee1d393052b3d573e754f3a1ac0

```bash
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/kubelet/prober
cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
            │    old.txt    │               new.txt               │
            │    sec/op     │   sec/op     vs base                │
HTTPProbe-8   142.71m ± 10%   33.33m ± 5%  -76.65% (p=0.000 n=10)

            │    old.txt    │               new.txt                │
            │     B/op      │     B/op      vs base                │
HTTPProbe-8   251.43Mi ± 0%   35.01Mi ± 0%  -86.08% (p=0.000 n=10)

            │   old.txt    │               new.txt               │
            │  allocs/op   │  allocs/op   vs base                │
HTTPProbe-8   3367.7k ± 0%   607.0k ± 0%  -81.98% (p=0.000 n=10)
```

aojea commented:
> the code doing this benchmark: https://github.com/kubernetes/kubernetes/commit/4c0578c9c8e19ee1d393052b3d573e754f3a1ac0

yeah, but add those results  to the commit message and/or as a comment on top of the benchmark function, so we have a place where we can look in the code

Taoup commented:
> > the code doing this benchmark: [4c0578c](https://github.com/kubernetes/kubernetes/commit/4c0578c9c8e19ee1d393052b3d573e754f3a1ac0)
> 
> yeah, but add those results to the commit message and/or as a comment on top of the benchmark function, so we have a place where we can look in the code

sry I misunderstood your intent, updated the commit.

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -142,9 +146,16 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
+		key := probeKey{podUID: pod.UID, containerName: container.Name, probeType: probeType}
+		reqAny, found := pb.httpProbeReqCache.Load(key)
+		req, _ := reqAny.(*http.Request)
+		if !found || (p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP) {
```
This is not very obvious (at least to me). Can you add a comment explaining what you're doing here and why?

Taoup replied:
I added this logic because I knew from many places that a pod is mutable.

As https://github.com/kubernetes/kubernetes/issues/108281#issuecomment-1058503524 states,  
> I don't think we are really saying that the IP can change underneath a running pod.
I'm considering removing this logic totally. If IP for a running pod won't change, can I assume the http probe request for a running pod won't stale?

Taoup committed:
cache http probe request in httpprober

cache stored in httpprober, using url as cache key

goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/kubelet/prober
cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
            │    old.txt    │       new-httpprober-cache.txt       │
            │    sec/op     │    sec/op     vs base                │
HTTPProbe-8   142.71m ± 10%   54.94m ± 13%  -61.51% (p=0.000 n=10)

            │    old.txt    │       new-httpprober-cache.txt       │
            │     B/op      │     B/op      vs base                │
HTTPProbe-8   251.43Mi ± 0%   79.11Mi ± 0%  -68.53% (p=0.000 n=10)

            │   old.txt   │      new-httpprober-cache.txt       │
            │  allocs/op  │  allocs/op   vs base                │
HTTPProbe-8   3.368M ± 0%   1.585M ± 0%  -52.94% (p=0.000 n=10)

bart0sh commented on pkg/probe/http/http.go:
```diff
@@ -90,7 +116,7 @@ type GetHTTPInterface interface {
 // If the HTTP response code is unsuccessful or HTTP communication fails, it returns Failure.
 // This is exported because some other packages may want to do direct HTTP probes.
-func DoHTTPProbe(req *http.Request, client GetHTTPInterface) (probe.Result, string, error) {
+var DoHTTPProbe = func(req *http.Request, client GetHTTPInterface) (probe.Result, string, error) {
```
Can you explain why is this change needed?

Taoup replied:
I changed this for the convenience of running the benchmark.  In the benchmark, this func variable is changed to a function always return true to mock the behavior of doing an actual http probe.  If this pr is to be merged, I think this change and the associated benchmark code is not necessary to be included.

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -142,7 +145,7 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
 
 	case p.HTTPGet != nil:
-		req, err := httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
+		req, err := pb.http.(*httpprobe.HttpProber).GetRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
```
Can this casting be avoided? It makes the code less clear.

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -49,6 +50,8 @@ type prober struct {
 
 	recorder record.EventRecorder
+
+	httpProbeReqCache sync.Map
```
This should be moved to the http probe code, I believe.

Taoup replied:
sure

bart0sh commented on pkg/probe/http/http.go:
```diff
@@ -58,21 +60,22 @@ func NewWithTLSConfig(config *tls.Config, followNonLocalRedirects bool) Prober {
+type HttpProber struct {
 	transport               *http.Transport
 	followNonLocalRedirects bool
+	httpRequestCache        *sync.Map
```
Why this is not enough and the same field is added to the `prober` struct?

Taoup replied:
> Why this is not enough and the same field is added to the `prober` struct?

Oh, after a bunch of rebase things, I forgot the original one is still in `probe` struct.  Will delete it.
